src/seraphis_wallet/address_utils.cpp

```diff
@@ -0,0 +1,92 @@
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// paired header
+#include "address_utils.h"
+
+#include "misc_log_ex.h"
+#include "seraphis_impl/jamtis_address_checksum.h"
+#include "seraphis_impl/serialization_demo_types.h"
+#include "seraphis_impl/serialization_demo_utils.h"
+
+// local headers
+
+// third party headers
+
+// standard headers
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "seraphis_wallet"
+
+using namespace sp::serialization;
+using namespace std;
+
+void get_destination_from_str(const std::string &address, JamtisDestinationV1 &dest_out)
+{
+    // 1. get checksum
+    std::string main_address = address.substr(6, address.length() - 14);
+    std::string checksum     = address.substr(address.length() - 8);
+    std::string checksum_calculated{sp::jamtis::create_address_checksum(address.substr(0, address.length() - 8))};
+
+    // 2. test checksum
+    CHECK_AND_ASSERT_THROW_MES(checksum == checksum_calculated, "Checksum mismatch at get_destination_from_str.");
+
+    // 3. prepare to recover JamtisDestinationV1
+    std::string serialized_address;
+    serialized_address = base32::decode(main_address);
+    sp::serialization::ser_JamtisDestinationV1 serializable_destination_recovered;
+    sp::serialization::try_get_serializable(
+        epee::strspan<std::uint8_t>(serialized_address), serializable_destination_recovered);
+
+    // 4. get destination
+    sp::serialization::recover_sp_destination_v1(serializable_destination_recovered, dest_out);
+}
+//-----------------------------------------------------------------
+void get_str_from_destination(const JamtisDestinationV1 &dest,
+    const JamtisAddressVersion address_version,
+    const JamtisAddressNetwork address_network,
+    std::string &address_out)
+{
+    // 1. initial fixed parameters
+    const std::string address_prefix = "xmra";
+
+    // 2. prepare to encode string
+    sp::serialization::ser_JamtisDestinationV1 serializable_destination;
+    make_serializable_sp_destination_v1(dest, serializable_destination);
+    std::string serialized_address;
+    sp::serialization::try_append_serializable(serializable_destination, serialized_address);
+
+    // 3. encode string and add to address
+    std::string address = address_prefix;
+    address += static_cast<char>(address_version);
+    address += static_cast<char>(address_network);
+    address += base32::encode(serialized_address);
```

Comment:
It was decided to switch to base32 for Jamtis? Interesting, I don't recall a discussion about it though.

Reply:
I think @tevador proposed a Base32 scheme together with a custom alphabet [discussed here](https://gist.github.com/tevador/50160d160d24cfc6c52ae02eb3d17024?permalink_comment_id=4399652#gistcomment-4399652) as integral part of his Jamtis spec, see [here](https://gist.github.com/tevador/50160d160d24cfc6c52ae02eb3d17024#35-base32-encoding). There was some more discussion; e.g. @hyc asked why not Base64 [here](https://gist.github.com/tevador/50160d160d24cfc6c52ae02eb3d17024?permalink_comment_id=4401848#gistcomment-4401848).

I had a chapter *The Case for Base32* in [my Reddit post](https://www.reddit.com/r/Monero/comments/zl4615/why_seraphis_jamtis_addresses_will_be_so_awfully/) about Jamtis addresses.

You could say that nobody opposed enough or came up with solid counter-arguments to topple Tevador's design decision, so we go along.